gc/2021/GC-07-27.md

```diff
@@ -0,0 +1,481 @@
+
+CW: in the past we have considered a virtue to keep Wasm small, push this kind of logic into a level above us, here, push into toolchain. To replicate C’s linking semantics, we pushed all logic into tools and tools-generated JS. Maybe in the future it is JS and module linking proposal. In other situations we would consider this a virtue, deferring how to manage types into toolchain.
+
+AR: don’t think so, this is a genuine part of the semantics, what is type equivalence. Outsourcing a central part of the semantics seems not such a great idea. We have made different decisions in different cases, whether it makes sense to put things into engines or toolchains. We haven’t been totally consistent. Driven by other considerations.
+
+TL: as tool author, prefer to keep tools simple. Our philosophy has been, we are pushing as much as cost and complexity into tools. Conflicting here, want Wasm engines to be simple, security risk there. When we have structural types, all this sharing, it is a lot simpler for producers. The performance cost of doing equirecursive cost on engines is too high.
+
+AR: consciously not talked about equirecursive, that is an orthogonal question
+
+CW: only if we have an alternative proposal
+
+AR: next part of the presentation
+
+BT: primarily think about engine complexity, moving this into userspace seems really complicated, considering all the layers, it will be even more complicated if this is not in the engine.
+
+AR: if we don’t need equirecursion, what we put in engine is not complicated, really just hash consing.
+
+CW: complexity of doing something at engine level that supports all lang that wants to compile to Wasm, will be more than any toolchain that wants to implement. Equi-canon is what engines will need to implement.
+
+AR: structural recursion, not equi
+
+RT: many examples where structural types are more convenient for languages, but for them we can use nominal types too, with some effort. Other problems, the flipside isn’t true.
+
+AR: none of these claims are true
+
+CW: should shut down this conversation, bounded quant is post-post-post MVP
+
+BT: meta-level - need to align on shipping something that people will use. Many things down the road we can think about, and many iterations. We need to actually make progress forward.
+
+RT: need to figure out the overhead making 5x slower
+
+CW: mvp will have so many casts that it will be slow
+
+FM: no point shipping something 5x slower
+
+TL: recent perf numbers are better than 5x slower, not as bad. Agree to align on shipping something
+
+RT: kotlin tried to do separate compilation and couldn’t get it to work
+
+AR: yea could be true, but can have many reasons for that, doesn’t mean we should give up on separate compilation
+
+RT: the issues were how MVP expresses and composes types, insufficient for them
+
+BT: when you use type across module boundary, you can’t refer to fields, that’s missing
+
+RT: importing bounded types…
```

Comment:
@RossTate do you remember the rest of what you were saying here?

Reply:
I think I was pointing out that importing types with bounds in order to get access to their fields does not make for a great separate compilation story. Instead, I had found that importing (unbounded) types and then importing field members for those types made for much simpler separate compilation.